aws/resource_aws_api_gateway_vpc_link.go

```diff
@@ -0,0 +1,200 @@
+	conn := meta.(*AWSClient).apigateway
+
+	input := &apigateway.GetVpcLinkInput{
+		VpcLinkId: aws.String(d.Id()),
+	}
+
+	resp, err := conn.GetVpcLink(input)
+	if err != nil {
+		if isAWSErr(err, apigateway.ErrCodeNotFoundException, "") {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	d.Set("name", resp.Name)
+	d.Set("description", resp.Description)
+	d.Set("target_arn", resp.TargetArns[0])
+	return nil
+}
+
+func resourceAwsApiGatewayVpcLinkUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).apigateway
+
+	operations := make([]*apigateway.PatchOperation, 0)
+
+	if d.HasChange("name") {
+		operations = append(operations, &apigateway.PatchOperation{
+			Op:    aws.String("replace"),
+			Path:  aws.String("/name"),
+			Value: aws.String(d.Get("name").(string)),
+		})
+	}
+
+	if d.HasChange("description") {
+		operations = append(operations, &apigateway.PatchOperation{
+			Op:    aws.String("replace"),
+			Path:  aws.String("/description"),
+			Value: aws.String(d.Get("description").(string)),
+		})
+	}
+
+	input := &apigateway.UpdateVpcLinkInput{
+		VpcLinkId:       aws.String(d.Id()),
+		PatchOperations: operations,
+	}
+
+	_, err := conn.UpdateVpcLink(input)
+	if err != nil {
+		if isAWSErr(err, apigateway.ErrCodeNotFoundException, "") {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{apigateway.VpcLinkStatusPending},
+		Target:     []string{apigateway.VpcLinkStatusAvailable},
+		Refresh:    apigatewayVpcLinkRefreshStatusFunc(conn, d.Id()),
+		Timeout:    8 * time.Minute,
+		MinTimeout: 3 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("[WARN] Error waiting for APIGateway Vpc Link status to be \"%s\": %s", apigateway.VpcLinkStatusAvailable, err)
+	}
+
+	return nil
+}
+
+func resourceAwsApiGatewayVpcLinkDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).apigateway
+
+	input := &apigateway.DeleteVpcLinkInput{
+		VpcLinkId: aws.String(d.Id()),
+	}
+
+	_, err := conn.DeleteVpcLink(input)
+	if err != nil {
+		if isAWSErr(err, apigateway.ErrCodeNotFoundException, "") {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	stateConf := resource.StateChangeConf{
+		Pending: []string{apigateway.VpcLinkStatusPending,
+			apigateway.VpcLinkStatusAvailable,
+			apigateway.VpcLinkStatusDeleting},
+		Target:     []string{"deleted"},
+		Timeout:    5 * time.Minute,
+		MinTimeout: 1 * time.Second,
+		Refresh: func() (interface{}, string, error) {
+			resp, err := conn.GetVpcLink(&apigateway.GetVpcLinkInput{
+				VpcLinkId: aws.String(d.Id()),
+			})
+			if err != nil {
+				if isAWSErr(err, apigateway.ErrCodeNotFoundException, "") {
+					return 1, "deleted", nil
+				}
+				return nil, "failed", err
+			}
+			return resp, *resp.Status, nil
+		},
+	}
+
+	if _, err := stateConf.WaitForState(); err != nil {
+		return err
+	}
+
+	d.SetId("")
```

Comment:
Nitpick: I know we still have it in many resources, but it's redundant. The ID is emptied automatically in the core.

See https://github.com/hashicorp/terraform/blob/096847c9f774bfb946de7413260b30f9f6041241/helper/schema/resource.go#L197-L206